.circleci/config.yml

```diff
@@ -168,6 +173,74 @@ commands:
                 git diff-tree --no-commit-id --name-only -r "$commit" -- "*Dockerfile";
               done
             )' >> $BASH_ENV
+  fail_if_debian_buster_docker_base_images_are_not_tagged_with_sha256:
+    steps:
+      - run:
+          name: Verify all debian buster images have sha256s.
+          when: always
+          command: |
+            set -e
+            # We are using debian for our base images.   Get all debain tags in FROM lines, strip the "AS .*" part
+            ALL_BUSTER_TAGS=`grep --include=\*Dockerfile -r '.' -e 'FROM debian:' | sed 's/[ ]*[a|A][s|S] .*$//g'`
+            while IFS= read -r line ; do
+              # if the line does not end in a git tag, echo the line and mark the build as bad.
+              if [[ ! $line =~ '@sha256:' ]]; then
+                echo "Give this docker image FROM line a sha256 in it's tag.   Failing the build."
+                echo $line;
+                SHOULD_FAIL="true";
+              fi
+            done \<<< "$ALL_BUSTER_TAGS"
+            if [[ "$SHOULD_FAIL" == "true" ]]; then
+              exit 1
+            fi
+  halt_if_no_relevant_files_changed_since_last_evaluation:
+    description: |
+      halts the job if no relevant files have changed.
+      This is determined by using a cached file containing the last git revision the build evaluated.
+      Branch is part of the cache key name, following the parameter, cache_key_part.
+      The build did not necessarily need to proceed on the last evaluation if no relevant files had changed,
+      and cache keys will still be updated on each evaluation before halting.
+      Relevant files are looked up from an exist file on the execution env, passed in as a parameter, the
+      file of relevant git files.
+    parameters:
+      cache_key_part:
+        description: part of the cache key name used by circle in the job.
+        type: string
+      file_of_relevant_git_files:
+        description: file to be provided to these steps, of relevant git files.
+        type: string
+    steps:
+      - restore_cache:
```

Comment:
The detect-and-determine solution (where it compares HEAD against origin) in the previous commit in this PR was more elegant and portable. Perhaps I didn't see the trade-off with using the circleci built-in cache management.

Reply:
there are two implementations that I pulled apart -- on for updates to master/release branchs. which this is --  there is no remote origin to compare to in that instance -- you must use a cache key as the build wont know which commits on the branch are new.   The other mechanism only works on prs.   I'm currently writing a shell script to merge the approaches, decoupled from circle.